app/src/main/java/de/rochefort/childmonitor/ListenActivity.kt

```diff
@@ -0,0 +1,122 @@
+                    Toast.LENGTH_SHORT).show()
+            val connectedText = findViewById<TextView>(R.id.connectedTo)
+            connectedText.text = bs.childDeviceName
+            val volumeView = findViewById<VolumeView>(R.id.volume)
+            volumeView.setVolumeHistory(bs.volumeHistory)
+            bs.setUpdateCallback { volumeView.postInvalidate() }
+            bs.setErrorCallback { postErrorMessage() }
+        }
+
+        override fun onServiceDisconnected(className: ComponentName) {
+            // This is called when the connection with the service has been
+            // unexpectedly disconnected -- that is, its process crashed.
+            // Because it is running in our same process, we should never
+            // see this happen.
+            Toast.makeText(this@ListenActivity, R.string.disconnected,
+                    Toast.LENGTH_SHORT).show()
+        }
+    }
+
+    private fun ensureServiceRunningAndBind(bundle: Bundle?) {
+        val context: Context = this
+        val intent = Intent(context, ListenService::class.java)
+        if (bundle != null) {
+            intent.putExtra("name", bundle.getString("name"))
+            intent.putExtra("address", bundle.getString("address"))
+            intent.putExtra("port", bundle.getInt("port"))
+            ContextCompat.startForegroundService(context, intent)
+        }
+        // Attempts to establish a connection with the service.  We use an
+        // explicit class name because we want a specific service
+        // implementation that we know will be running in our own process
+        // (and thus won't be supporting component replacement by other
+        // applications).
+        if (bindService(intent, connection, BIND_AUTO_CREATE)) {
+            this.shouldUnbind = true
+            Log.i(TAG, "Bound listen service")
+        } else {
+            Log.e(TAG, "Error: The requested service doesn't " +
+                    "exist, or this client isn't allowed access to it.")
+        }
+    }
+
+    private fun doUnbindAndStopService() {
+        if (this.shouldUnbind) {
+            // Release information about the service's state.
+            unbindService(connection)
+            this.shouldUnbind = false
+        }
+        val context: Context = this
+        val intent = Intent(context, ListenService::class.java)
+        context.stopService(intent)
+    }
+
+    fun postErrorMessage() {
+        val status = findViewById<TextView>(R.id.textStatus)
+        status.post { status.setText(R.string.disconnected) }
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        val bundle = this.intent.extras
+        ensureServiceRunningAndBind(bundle)
+        this.volumeControlStream = AudioManager.STREAM_MUSIC
+        setContentView(R.layout.activity_listen)
+        val statusText = findViewById<TextView>(R.id.textStatus)
+        statusText.setText(R.string.listening)
```

Comment:
findViewById does not promise to return non null value, so maybe we should use the ? operator here?

Reply:
While the function does not guarantee not to return null, it would only return null, if the resources do not match the code. And then the question is, how to handle that?
I think, simply crashing with a nice stack-trace would be preferable over continuing to work but having some fields unset.

Probably, the pattern in the android documentation should be followed:
https://developer.android.com/kotlin/common-patterns#fragment-nullability

Reply:
I agree with the logic that if this returns null we are looking at a bug so it's okay to crash.
Lets leave it as is, the lateinit stuff does not apply here since this is a local variable.